tests/macros/utils.js

```diff
@@ -4,83 +4,269 @@
 
 // Provides utilities that as a whole constitute the macro test framework.
 const { execSync } = require('child_process');
+const fs = require('fs');
 const os = require('os');
+const path = require('path');
 
 const vnu = require('vnu-jar');
 
 const Environment = require('../../src/environment.js');
 const Templates = require('../../src/templates.js');
 
-// When we were doing mocha testing, we used this.macro to hold this.
-// But Jest doesn't use the this object, so we just store the object here.
+/**
+ * @typedef Macro
+ * @property {typeof Environment.prototype.prototypeEnvironment} ctx
+ *           Give the test-case writer access to the macro's globals (ctx).
+ *           For example, "macro.ctx.env.locale" can be manipulated to something
+ *           other than 'en-US' or "macro.ctx.wiki.getPage" can be mocked
+ *           using `jest.fn()` to avoid network calls.
+ *
+ * @property {(name: string, result: string) => void} mockTemplate
+ *           When writing tests for a macro that invokes other macros with
+ *           the `template()` function, you sometimes want to specify
+ *           a mock return value for those other macros.
+ *
+ *           This function provides a much easier way to handle that than
+ *           using `jest.fn()` directly.
+ *
+ *           To unmock a template result, simply call `unmockTemplate()`
+ *           with the same `name`.
+ *
+ * @property {(name: string) => boolean} unmockTemplate
+ *           Stops mocking the result of a `template()` function call.
+ *
+ *           **Returns:**
+ *           - `true` if a macro result has previously been
+ *             mocked using `mockTemplate()`.
+ *           - `false` otherwise.
+ *
+ * @property {(...args: any[]) => Promise<string>} call
+ *           Use this function to make test calls on the named macro, if applicable.
+ *           Its arguments become the arguments to the macro. It returns a promise.
+ */
+
+/**
+ * When we were doing mocha testing, we used this.macro to hold this.
+ * But Jest doesn't use the this object, so we just store the object here.
+ * @type {Macro}
+ */
 let macro = null;
 
+/**
+ * Asserts that value is truthy.
+ *
+ * @param {any} x The value.
+ */
 function assert(x) {
-    expect(x).toBe(true);
+    expect(x).toBeTruthy();
 }
 
+/**
+ * Asserts deep equality of actual and expected.
+ *
+ * @param {any} x Actual value.
+ * @param {any} y Potential expected value.
+ */
 assert.equal = (x, y) => {
     expect(x).toEqual(y);
 };
 
 assert.eventually = {
+    /**
+     * Asserts deep equality of actual and expected.
+     *
+     * @param {any|Promise<any>} x A promise which resolves to the actual value.
+     * @param {any} y Potential expected value.
+     */
     async equal(x, y) {
         expect(await x).toEqual(y);
     }
 };
 
+/**
+ * Asserts that list includes the supplied element.
+ *
+ * @template T Type of values in list.
```

Comment:
The elements of the list don't all have to be the same type, so maybe better here to use any instead of T?

Reply:
I’ve set the list to&nbsp;`string|any[]`, but the&nbsp;`element` has&nbsp;to&nbsp;be&nbsp;`string|T`, otherwise the result would collapse into&nbsp;`any`.